modules/Installer/AWS.Tools.Installer.psm1

```diff
@@ -501,75 +463,127 @@ function Install-AWSToolsModule {
-                    $saveAWSToolsModuleParams = @{
-                        RequiredVersion        = $versionToInstall
-                        TemporaryRepoDirectory = $temporaryRepoDirectory
-                        SavedModules           = $savedModules
-                        Proxy                  = $Proxy
-                        ProxyCredential        = $ProxyCredential
-                        Confirm                = $false
-                        WhatIf                 = $false
+                    try {
+                        $httpClientHandler = [System.Net.Http.HttpClientHandler]::new()
+                        if ($Proxy) {
+                            $httpClientHandler.Proxy = [System.Net.WebProxy]::new($Proxy)
+                            if ($ProxyCredential) {
+                                $httpClientHandler.Proxy.Credentials = $ProxyCredential.GetNetworkCredential()
+                            }
+                        }
+                        $httpClient = [System.Net.Http.HttpClient]::new($httpClientHandler);
+
+                        Add-Type $script:ParallelDownloaderClassCode -ReferencedAssemblies System.Net.Http
+                        [ParallelDownloader]$parallelDownloader = [ParallelDownloader]::new($httpClient)
+
+                        [string[]]$modulesToDownload = $modulesToInstall
+                        [System.Collections.Generic.HashSet[string]]$savedModules = New-Object -TypeName System.Collections.Generic.HashSet[string]
+
+                        Write-Verbose "[$($MyInvocation.MyCommand)] Downloading modules ($modulesToDownload)"
+
+                        while ($modulesToDownload) {
+                            [string[]]$dependencies = @()
+
+                            $tasks = $modulesToDownload | Where-Object { $savedModules.Add($_) } | ForEach-Object {
+                                [string]$nupkgFilePath = Join-Path $temporaryRepoDirectory "$($_).$($RequiredVersion).nupkg"
+                                Write-Verbose "[$($MyInvocation.MyCommand)] Downloading module $($_) to $TemporaryRepoDirectory"
+                                [System.Threading.Tasks.Task]$task = $parallelDownloader.DownloadToFile("https://www.powershellgallery.com/api/v2/package/$_/$RequiredVersion", $nupkgFilePath)
+                                @{
+                                    Task       = $task
+                                    ModuleName = $_
+                                    Path       = $nupkgFilePath
+                                }
+                            }
+                            $tasks | ForEach-Object {
+                                $_.Task.Wait()
+                                $dependencies += Get-AWSToolsModuleDependenciesAndValidate -Path $_.Path -Name $_.ModuleName
+                            }
+
+                            $modulesToDownload = $dependencies | Sort-Object -Unique
+                        }
```

Comment:
Since we are already using a C# class for this, why don't we just pass the entire list and get the parallelization code out of powershell?

Reply:
1) I don't want to have too complex code in C# because it is provided as a `string`, so you lose a lot of readability by transferring logic to C#.
2) This allows me to start checking signatures while the downloads are underway
3) I have now updated the PR to support cancellation of the tasks in case of error